topology/nhlt/intel/ssp-nhlt.c

```diff
@@ -0,0 +1,306 @@
+	const char *bclk_invert = NULL;
+	const char *fsync = NULL;
+	const char *fsync_invert = NULL;
+	long mclk_freq = 0;
+	long bclk_freq = 0;
+	long fsync_freq = 0;
+	long tdm_slots = 0;
+	long tdm_slot_width = 0;
+	long tx_slots = 0;
+	long rx_slots = 0;
+	long ret;
+
+	struct dai_values ssp_hw_data[] = {
+		{"format", SND_CONFIG_TYPE_STRING, NULL, NULL, &format},
+		{"mclk", SND_CONFIG_TYPE_STRING, NULL, NULL, &mclk},
+		{"bclk", SND_CONFIG_TYPE_STRING, NULL, NULL, &bclk},
+		{"fsync", SND_CONFIG_TYPE_STRING, NULL, NULL, &fsync},
+		{"bclk_invert", SND_CONFIG_TYPE_STRING, NULL, NULL, &bclk_invert},
+		{"fsync_invert", SND_CONFIG_TYPE_STRING, NULL, NULL, &fsync_invert},
+		{"fsync_freq", SND_CONFIG_TYPE_INTEGER, NULL, &fsync_freq, NULL},
+		{"bclk_freq", SND_CONFIG_TYPE_INTEGER, NULL, &bclk_freq, NULL},
+		{"mclk_freq", SND_CONFIG_TYPE_INTEGER, NULL, &mclk_freq, NULL},
+		{"tdm_slots", SND_CONFIG_TYPE_INTEGER, NULL, &tdm_slots, NULL},
+		{"tdm_slot_width", SND_CONFIG_TYPE_INTEGER, NULL, &tdm_slot_width, NULL},
+		{"tx_slots", SND_CONFIG_TYPE_INTEGER, NULL, &tx_slots, NULL},
+		{"rx_slots", SND_CONFIG_TYPE_INTEGER, NULL, &rx_slots, NULL},
+	};
+
+	ret = find_set_values(&ssp_hw_data[0], ARRAY_SIZE(ssp_hw_data), cfg, top,
+			      "Class.Base.hw_config");
+	if (ret < 0)
+		return ret;
+
+	return ssp_hw_set_params(nhlt, format, mclk, bclk, bclk_invert, fsync, fsync_invert,
+				 mclk_freq, bclk_freq, fsync_freq, tdm_slots, tdm_slot_width,
+				 tx_slots, rx_slots);
+}
+
+/* init ssp parameters, should be called before parsing dais */
+int nhlt_ssp_init_params(struct intel_nhlt_params *nhlt)
+{
+	return ssp_init_params(nhlt);
+}
+
+int nhlt_ssp_get_ep_count(struct intel_nhlt_params *nhlt)
+{
+	return ssp_get_vendor_blob_count(nhlt);
+}
+
+int nhlt_ssp_get_ep(struct intel_nhlt_params *nhlt, struct endpoint_descriptor **eps,
+		    int dai_index)
+{
+	struct endpoint_descriptor ep;
+	struct ssp_device_specific_config ssp_conf;
+	struct formats_config f_conf;
+	struct format_config f_conf1[8];
+	uint32_t sample_rate;
+	uint16_t channel_count;
+	uint32_t bits_per_sample;
+	uint32_t virtualbus_id;
+	uint32_t formats_count;
+	uint8_t *ep_target;
+	size_t blob_size;
+	int ret;
+	int i;
+
+	/*
+	 * nhlt ssp structure:
+	 *
+	 * endpoint_descriptor, sizeof(struct endpoint_descriptor)
+	 * device_specific_config (headset), sizeof(struct ssp_device_specific_config)
+	 * formats_config (formats_count), sizeof(struct formats_config)
+	 * format_config (waveex), sizeof(struct format_config)
+	 * vendor_blob sizeof(vendor_blob)
+	 */
+
+	ret = ssp_get_params(nhlt, dai_index, &virtualbus_id, &formats_count);
+	if (ret < 0) {
+		fprintf(stderr, "nhlt_ssp_get_ep: ssp_get_params failed\n");
+		return ret;
+	}
+
+	ep.link_type = NHLT_LINK_TYPE_SSP;
+	ep.instance_id = 0;
+	ep.vendor_id = NHLT_VENDOR_ID_INTEL;
+	ep.device_id = NHLT_DEVICE_ID_INTEL_I2S_TDM;
+	ep.revision_id = 0;
+	ep.subsystem_id = 0;
+	ep.device_type = 0;
+	ep.direction = NHLT_ENDPOINT_DIRECTION_RENDER;
```

Comment:
what about capture?

Reply:
so not sure what to do here... NHLT doesn't have duplex definition, so should I just duplicate the endpoint data in that case? anyway can take this somehow from direction param

Reply:
Yes I think the endpoints are duplicated for the SSP. That's what you see from the NHTL tool from Bob Moore.

Reply:
sure I can add the duplication